core/runtime/src/main/java/io/quarkiverse/flow/providers/CredentialsProviderSecretManager.java

```diff
@@ -0,0 +1,153 @@
+            singleProvider = Optional.of(providers.get());
+        }
+
+        for (CredentialsProvider p : all) {
+            Named n = p.getClass().getAnnotation(Named.class);
+            if (n != null && !n.value().isBlank()) {
+                providerBeanCache.putIfAbsent(n.value(), p);
+            }
+        }
+    }
+
+    /**
+     * Recorder checks this to decide whether to wire this SM or let the SDK fallback handle resolution.
+     */
+    public boolean hasAnyProviders() {
+        return hasAnyProviders;
+    }
+
+    @Override
+    public Map<String, String> secret(String secretName) {
+        CredentialsProvider provider = secretCache.computeIfAbsent(secretName, this::resolveProviderForSecret);
+        Map<String, String> creds = provider.getCredentials(secretName);
+        if (creds == null || creds.isEmpty()) {
+            throw new IllegalStateException("No credentials found for secret '" + secretName
+                    + "' from provider @" + namedOf(provider) + " (" + provider.getClass().getName() + ")");
+        }
+        return creds;
+    }
+
+    private CredentialsProvider resolveProviderForSecret(String secretName) {
+        String per = nullSafeTrim(flowSecrets.credentialsProviderNames().get(secretName));
+        if (per != null) {
+            return selectConfigProvider(per);
+        }
+
+        Optional<String> global = flowSecrets.credentialsProviderName().map(String::trim).filter(s -> !s.isEmpty());
+        if (global.isPresent()) {
+            return selectConfigProvider(global.get());
+        }
+
+        if (singleProvider.isPresent()) {
+            return singleProvider.get();
+        }
+
+        throw new IllegalStateException(
+                "Multiple or no CredentialsProvider beans found and no selection provided. " +
+                        "Set either '" + GLOBAL_KEY + "' or '" + PER_SECRET_KEY.replace("<secret>", secretName) + "'. " +
+                        "Available: " + availableProviders());
+    }
+
+    /**
+     * Centralized config → provider resolution with cache + CDI fallback.
+     */
+    private CredentialsProvider selectConfigProvider(String named) {
+        CredentialsProvider cached = providerBeanCache.get(named);
+        if (cached != null)
+            return cached;
+
+        Instance<CredentialsProvider> handle = providers.select(NamedLiteral.of(named));
+        if (handle.isResolvable() && !handle.isAmbiguous()) {
+            CredentialsProvider p = handle.get();
+            providerBeanCache.putIfAbsent(named, p);
+            return p;
+        }
+        if (handle.isAmbiguous()) {
+            throw new IllegalStateException("Multiple CredentialsProvider beans match @Named='" + named
+                    + "'. Available: " + availableProviders());
+        }
+        throw new IllegalStateException("CredentialsProvider @Named='" + named
+                + "' not found. Available: " + availableProviders());
```

Comment:
You have already populated the providedBeanCache will all beans at startup.  I do not think more instances will be added after that to `providers`, so accesing the map (which does not need to be concurrent in that case) will be enought
But if more instances might be added after startup, then, rather than loading them at startup, then lets use computeIfAbsent as you are doing for the secret cache and avoid a get followed by a put.

Reply:
I'd rather populate it at this bean startup before the application runs, but at init. In case the CDI changes (like in devmode or a config refresh), we change the cache. 99% of the time, we will return the cache on workflow runs.